File: cs/cslox/Parser/AstPrinter.cs
using System;
using System.Text;
using cslox.Scanning;

namespace cslox.Parser;

class AstPrinter : IExprVisitor<string>
{
    public string Print(Expr expr)
    {
        return expr.Accept(this);
    }

    public string VisitAssignExpr(Assign expr)
    {
        return Parenthesize2("=", expr.Name.Lexeme, expr.Value);
    }

    public string VisitBinaryExpr(Binary expr)
    {
        return Parenthesize(expr.Operator.Lexeme, expr.Left, expr.Right);
    }

    public string VisitCallExpr(Call expr)
    {
        return Parenthesize2("call", expr.Callee, expr.Arguments);
    }

    public string VisitGetExpr(Get expr)
    {
        return Parenthesize2(".", expr.Object, expr.Name.Lexeme);
    }

    public string VisitGroupingExpr(Grouping expr)
    {
        return Parenthesize("group", expr.Expression);
    }

    public string VisitLogicalExpr(Logical expr)
    {
        return Parenthesize(expr.Operator.Lexeme, expr.Left, expr.Right);
    }

    public string VisitSetExpr(Set expr)
    {
        return Parenthesize2("=", expr.Object, expr.Name.Lexeme, expr.Value);
    }

    public string VisitSuperExpr(Super expr)
    {
        return Parenthesize2("super", expr.Method);
    }

    public string VisitLiteralExpr(Literal expr)
    {
        if (expr.Value == null)
        {
            return "nil";
        }
        return expr.Value.ToString()!;
    }

    public string VisitThisExpr(This expr)
    {
        return "this";
    }

    public string VisitUnaryExpr(Unary expr)
    {
        return Parenthesize(expr.Operator.Lexeme, expr.Right);
    }

    public string VisitVariableExpr(Variable expr)
    {
        return expr.Name.Lexeme;
    }

    private string Parenthesize(string name, params Expr[] exprs)
    {
        var builder = new StringBuilder();
        builder.Append("(").Append(name);
        foreach (var expr in exprs)
        {
            builder.Append(" ");
            builder.Append(expr.Accept(this));
        }
        builder.Append(")");

        return builder.ToString();
    }

    private string Parenthesize2(string name, params object[] parts)
    {
        var builder = new StringBuilder();
        builder.Append("(").Append(name);
        Transform(builder, parts);
        builder.Append(")");

        return builder.ToString();
    }

    private void Transform(StringBuilder builder, params object[] parts)
    {
        foreach (var part in parts)
        {
            builder.Append(" ");
            if (part is Expr)
            {
                builder.Append(((Expr)part).Accept(this));
            }
            else if (part is Stmt)
            {
                // builder.Append(((Stmt)part).Accept<string>(this));
            }
            else if (part is Token)
            {
                builder.Append(((Token)part).Lexeme);
            }
            //else if (part is List)
            //{
            //    builder.Append((List))part).ToArray());
            //}
            else
            {
                builder.Append(part);
            }
        }
    }
}

File: tree-walk/go/internal/lox/ast_printer.go
package lox

import (
	"fmt"
	"strings"
)

type AstPrinter struct{}

func (p *AstPrinter) Print(expression Expr) string {
	return expression.Accept(p).(string)
}

func (p *AstPrinter) VisitAssignExpr(expr *Assign) any {
	return p.parenthesize2("=", expr.Name.Lexeme, expr.Value)
}

func (p *AstPrinter) VisitBinaryExpr(expr *Binary) any {
	return p.parenthesize(expr.Operator.Lexeme, expr.Left, expr.Right)
}

func (p *AstPrinter) VisitCallExpr(expr *Call) any {
	return p.parenthesize2("call", expr.Callee, expr.Arguments)
}

func (p *AstPrinter) VisitGetExpr(expr *Get) any {
	return p.parenthesize2(".", expr.Object, expr.Name.Lexeme)
}

func (p *AstPrinter) VisitGroupingExpr(expr *Grouping) any {
	return p.parenthesize("group", expr.Expression)
}

func (p *AstPrinter) VisitLiteralExpr(expr *Literal) any {
	if expr.Value == nil {
		return "nil"
	}

	return fmt.Sprintf("%v", expr.Value)
}

func (p *AstPrinter) VisitLogicalExpr(expr *Logical) any {
	return p.parenthesize(expr.Operator.Lexeme, expr.Left, expr.Right)
}

func (p *AstPrinter) VisitSetExpr(expr *Set) any {
	return p.parenthesize2("=", expr.Object, expr.Name.Lexeme, expr.Value)
}

func (p *AstPrinter) VisitSuperExpr(expr *Super) any {
	return p.parenthesize2("super", expr.Method)
}

func (p *AstPrinter) VisitThisExpr(expr *This) any {
	return "this"
}

func (p *AstPrinter) VisitUnaryExpr(expr *Unary) any {
	return p.parenthesize(expr.Operator.Lexeme, expr.Right)
}

func (p *AstPrinter) VisitVariableExpr(expr *Variable) any {
	return expr.Name.Lexeme
}

func (p *AstPrinter) parenthesize(name string, exprs ...Expr) string {
	var builder strings.Builder
	builder.WriteString("(")
	builder.WriteString(name)
	for _, expr := range exprs {
		builder.WriteString(" ")
		v, _ := expr.Accept(p).(string)
		builder.WriteString(v)
	}
	builder.WriteString(")")

	return builder.String()
}

func (p *AstPrinter) parenthesize2(name string, parts ...any) string {
	var builder strings.Builder
	builder.WriteString("(")
	builder.WriteString(name)
	p.transform(builder, parts)
	builder.WriteString(")")

	return builder.String()
}

func (p *AstPrinter) transform(builder strings.Builder, parts ...any) {
	for _, part := range parts {
		builder.WriteString(" ")
		if expr, ok := part.(Expr); ok {
			v, _ := expr.Accept(p).(string)
			builder.WriteString(v)
		} else if _, ok := part.(Stmt); ok {
			// TODO
		} else if token, ok := part.(*Token); ok {
			builder.WriteString(token.Lexeme)
		} else {
			v, _ := part.(string)
			builder.WriteString(v)
		}
	}
}
